text/0000-impl-specialization.md

```diff
@@ -0,0 +1,1652 @@
+    default fn extend(iter: I) { .. }
+}
+
+impl<T> Vec<T> {
+    fn extend(slice: &[T]) { .. }
+}
+```
+
+This RFC proposes to permit such specialization at the inherent impl level. The
+semantics is defined in terms of the folklore desugaring above.
+
+(Note: this example was chosen purposefully: it's possible to use specialization
+at the inherent impl level to avoid refactoring the `Extend` trait as described
+in the Motivation section.)
+
+One tricky aspect here is that, since there is no explicit trait definition,
+there is no general signature that each definition of an inherent item must
+match. Thinking about `Vec` above, for example, notice that the two signatures
+for `extend` look superficially different, although it's clear that the first
+impl is the more general of the two.
+
+We propose a very simple-minded conceptual desugaring: each item desugars into a
+distinct trait, with type parameters for e.g. each argument and the return
+type. All concrete type information then emerges from desugaring into impl
+blocks. Thus, for example:
+
+```
+impl<T, I> Vec<T> where I: IntoIterator<Item = T> {
+    default fn extend(iter: I) { .. }
+}
+
+impl<T> Vec<T> {
+    fn extend(slice: &[T]) { .. }
+}
+
+// Desugars to:
+
+trait Vec_extend<Arg, Result> {
+    fn extend(Arg) -> Result;
+}
+
+impl<T, I> Vec_extend<I, ()> for Vec<T> where I: IntoIterator<Item = T> {
+    default fn extend(iter: I) { .. }
+}
+
+impl<T> Vec_extend<&[T], ()> for Vec<T> {
+    fn extend(slice: &[T]) { .. }
+}
+```
+
+All items of a given name must desugar to the same trait, which means that the
+number of arguments must be consistent across all impl blocks for a given `Self`
+type. In addition, we require that *all of the impl blocks overlap* (meaning
+that there is a single, most general impl). Without these constraints, we would
+implicitly be permitting full-blown overloading on both arity and type
+signatures. For the time being at least, we want to restrict overloading to
+explicit uses of the trait system, as it is today.
+
+This "desugaring" semantics has the benefits of allowing inherent item
+specialization, and also making it *actually* be the case that inherent impls
+are really just implicit traits -- unifying the two forms of dispatch. Note that
+this is a breaking change, since examples like the following are (surprisingly!)
+allowed today:
+
+```rust
+struct Foo<A, B>(A, B);
+
+impl<A> Foo<A,A> {
+    fn foo(&self, _: u32) {}
+}
+
+impl<A,B> Foo<A,B> {
+    fn foo(&self, _: bool) {}
+}
+
+fn use_foo<A, B>(f: Foo<A,B>) {
+    f.foo(true)
+}
+```
+
+As has been proposed
+[elsewhere](https://internals.rust-lang.org/t/pre-rfc-adjust-default-object-bounds/2199/),
+this "breaking change" would be made available through a feature flag that must
+be used even after stabilization (to opt in to specialization of inherent
+impls); the full details will depend on pending revisions to
+[RFC 1122](https://github.com/rust-lang/rfcs/pull/1122).
```

Comment:
This sort of thing is the obvious candidate for removal in a hypothetical Rust 2.0. Maybe we should start a comprehensive list of such changes.